lunabot_control/scripts/homing_controller.py

```diff
@@ -0,0 +1,361 @@
+        v_C_april = T_camera_april @ p_A_zero
+        # if (T_camera_april == compare):
+        #     self._effort_msg.left_drive = 0
+        #     self._effort_msg.right_drive = 0
+
+        v_1C[2] = 0
+        v_C_april[2] = 0
+        v_C_april[3] = 0
+
+        # Finding the error angle between the two points, v_1C and v_2C, using dot product (in radians)
+
+        # v_1C = v_1C[0:2]
+        # v_2C = v_2C[0:2]
+
+        angle = np.arctan2(
+            v_1C[0] * v_2C[1] - v_2C[0] * v_1C[1], v_1C[0] * v_1C[1] + v_2C[0] * v_2C[1]
+        )
+
+        angle += 3 * np.pi / 2 + 0.3
+        if angle > np.pi:
+            angle -= 2 * np.pi
+
+        # determining the linear translational error (in x and y) of the robot considered
+
+        self.v_C_april = np.array(
+            [
+                self.T_camera_april_msg.position.x,
+                self.T_camera_april_msg.position.z,
+                0,
+                0,
+            ]
+        )
+
+        print(v_C_april)
+        distance = np.linalg.norm(v_C_april)
+        # Replaced v_2c with 0,0,0,1 for the sim because the distance being calculated
+        # Was seemingly off from the apriltag (based on mapping out values, was 0,0,0)
+
+        # write PD controller here with the output being lin, ang velocity
+        # Define the K constants for P, I, and D
+        # For sim tuning, lin and ang are flipped.
+        # For Testing angular tuning only change the first value
+
+        # Define the errors
+        # lin_error_dist = np.sqrt((lin_error[0])**2 + (lin_error[1])**2)
+        self._curr_error = np.array(
+            [distance - self.linear_setpoint, angle - self.angular_setpoint]
+        )
+
+        print("self._curr_error: ", self._curr_error)
+        self._error_total += self._curr_error
+
+        # Computing PID errors
+        ctrl = self._curr_error * self.KP
+        ctrl += self._error_total * self.KI
+        ctrl += (self._curr_error - self._prev_error) * self.KD
+
+        # Set current error to previous error
+        self._prev_error = self._curr_error
+
+        # Converting errors into the lin_vel, ang_vel
+
+        twist = np.zeros(2)  # lin, ang velocity
+
+        # Setting thresholds to stop movement
+
+        if np.abs(self._curr_error[0]) < self.linear_threshold:
+            ctrl[0] = 0
+        if np.abs(self._curr_error[1]) < self.angular_threshold:
+            ctrl[1] = 0
+
+        twist[0] = -ctrl[0]
+        twist[1] = ctrl[1]
+
+        print("twist", twist)
+
+        """For real robot"""
+        ctrl = self.diff_drive_model(
+            twist
+        )  # computes wheel velocities from lin, ang vel (twist)
+
+        self._effort_msg.left_drive = self.constrain(ctrl[0])
+        self._effort_msg.right_drive = self.constrain(ctrl[1])
+
+        print("left: ", self.constrain(ctrl[0]))
+        print("right: ", self.constrain(ctrl[1]))
+
+        """ For simulation (lin and ang vel swapped)
```

Comment:
Question: did we ever figure out why this is the case and / or if we can fix this

Reply:
Only way to fix is to use a custom gazebo plugin and add the negative in there. could be worth it if we want to improve the simulator next year. For now, just will add a TODO

Reply:
Fixed in https://github.com/PurdueLunabotics/purdue_lunabotics/pull/43/commits/32117e5f9cf768840680971d8b954971afe24c81